# 기술통계 실습 1 (p. 214)
- 자동차 연비 Data Set에서 기술통계치 구하기

In [25]:
import pandas as pd 
import numpy as np
import matplotlib
from matplotlib import pyplot as plt

#통계 분석 패키지
from scipy import stats
from statsmodels.stats.proportion import proportions_ztest

In [26]:
df_mycars = pd.read_csv('/home/piai/다운로드/실습화일/1 Big Data 분석 기초/mycars.csv') #데이터 가져오기

In [27]:
df_mycars.head()

manufacturer model  displacement  year  cylinder automatic driving  mpg  \
0         audi    a4           1.8  1999         4      auto       f   18   
1         audi    a4           1.8  1999         4    manual       f   21   
2         audi    a4           2.0  2008         4    manual       f   20   
3         audi    a4           2.0  2008         4      auto       f   21   
4         audi    a4           2.8  1999         6      auto       f   16   

   highway_mileage fuel    class  
0               29    p  compact  
1               29    p  compact  
2               31    p  compact  
3               30    p  compact  
4               26    p  compact

In [28]:
df_mycars.describe().round(3) #각 열에 대한 주요 통계치를 요약하여 출력 (소수점 셋쨋자리까지 표시)

displacement     year  cylinder      mpg  highway_mileage
count       234.000   234.00   234.000  234.000          234.000
mean          3.472  2003.50     5.889   16.859           23.440
std           1.292     4.51     1.612    4.256            5.955
min           1.600  1999.00     4.000    9.000           12.000
25%           2.400  1999.00     4.000   14.000           18.000
50%           3.300  2003.50     6.000   17.000           24.000
75%           4.600  2008.00     8.000   19.000           27.000
max           7.000  2008.00     8.000   35.000           44.000

In [29]:
df = df_mycars[['driving','mpg']] # 시내에서 연비 통계치
df.groupby('driving').describe()

mpg                                                   
         count       mean       std   min   25%   50%   75%   max
driving                                                          
4        103.0  14.330097  2.874459   9.0  13.0  14.0  16.0  21.0
f        106.0  19.971698  3.626510  11.0  18.0  19.0  21.0  35.0
r         25.0  14.080000  2.215852  11.0  12.0  15.0  15.0  18.0

In [30]:
df = df_mycars[['driving','highway_mileage']] # 고속도로에서 연비 통계치
df.groupby('driving').describe()

highway_mileage                                                   
                  count       mean       std   min   25%   50%   75%   max
driving                                                                   
4                 103.0  19.174757  4.078704  12.0  17.0  18.0  22.0  28.0
f                 106.0  28.160377  4.206881  17.0  26.0  28.0  29.0  44.0
r                  25.0  21.000000  3.662877  15.0  17.0  21.0  24.0  26.0

# 기술통계 실습 2 (p. 217)
- 자동차의 모델과 연비(mpg)에 대한 자료를 조사한 Table이다.
- 이 범주형 변수인 구동방식, 모델에 대해 counts, percents, cumulative counts, cumulative percents를 계산하시오.

In [42]:
df_mycars = pd.read_csv('/home/piai/다운로드/실습화일/1 Big Data 분석 기초/mycars.csv') #데이터 가져오기

In [43]:
df = df_mycars[['driving', 'model']] #1. 모델 Table
df

driving   model
0         f      a4
1         f      a4
2         f      a4
3         f      a4
4         f      a4
..      ...     ...
229       f  passat
230       f  passat
231       f  passat
232       f  passat
233       f  passat

[234 rows x 2 columns]

In [44]:
count = df_mycars['model'].value_counts().sort_index()
count.head() #차례대로 5개만 출력

4runner 4wd    6
a4             7
a4 quattro     8
a6 quattro     3
altima         6
Name: model, dtype: int64

In [45]:
cumcnt = np.cumsum(count)
cumcnt.head()

4runner 4wd     6
a4             13
a4 quattro     21
a6 quattro     24
altima         30
Name: model, dtype: int64

In [46]:
percent = count/sum(count)*100
percent.head()

4runner 4wd    2.564103
a4             2.991453
a4 quattro     3.418803
a6 quattro     1.282051
altima         2.564103
Name: model, dtype: float64

In [47]:
cumpct = np.cumsum(percent)
cumpct.head()

4runner 4wd     2.564103
a4              5.555556
a4 quattro      8.974359
a6 quattro     10.256410
altima         12.820513
Name: model, dtype: float64

In [48]:
count_data = pd.DataFrame({'Count': count, 'CumCnt':cumcnt,'Percent':percent, 'CumPct':cumpct})
count_data.columns.name='model'
count_data

model                   Count  CumCnt   Percent      CumPct
4runner 4wd                 6       6  2.564103    2.564103
a4                          7      13  2.991453    5.555556
a4 quattro                  8      21  3.418803    8.974359
a6 quattro                  3      24  1.282051   10.256410
altima                      6      30  2.564103   12.820513
c1500 suburban 2wd          5      35  2.136752   14.957265
camry                       7      42  2.991453   17.948718
camry solara                7      49  2.991453   20.940171
caravan 2wd                11      60  4.700855   25.641026
civic                       9      69  3.846154   29.487179
corolla                     5      74  2.136752   31.623932
corvette                    5      79  2.136752   33.760684
dakota pickup 4wd           9      88  3.846154   37.606838
durango 4wd                 7      95  2.991453   40.598291
expedition 2wd              3      98  1.282051   41.880342
explorer 4wd                6     104  2.564103   44.444444
f150 pickup 4wd             7     111  2.991453   47.435897
forester awd                6     117  2.564103   50.000000
grand cherokee 4wd          8     125  3.418803   53.418803
grand prix                  5     130  2.136752   55.555556
gti                         5     135  2.136752   57.692308
impreza awd                 8     143  3.418803   61.111111
jetta                       9     152  3.846154   64.957265
k1500 tahoe 4wd             4     156  1.709402   66.666667
land cruiser wagon 4wd      2     158  0.854701   67.521368
malibu                      5     163  2.136752   69.658120
maxima                      3     166  1.282051   70.940171
mountaineer 4wd             4     170  1.709402   72.649573
mustang                     9     179  3.846154   76.495726
navigator 2wd               3     182  1.282051   77.777778
new beetle                  6     188  2.564103   80.341880
passat                      7     195  2.991453   83.333333
pathfinder 4wd              4     199  1.709402   85.042735
ram 1500 pickup 4wd        10     209  4.273504   89.316239
range rover                 4     213  1.709402   91.025641
sonata                      7     220  2.991453   94.017094
tiburon                     7     227  2.991453   97.008547
toyota tacoma 4wd           7     234  2.991453  100.000000

In [49]:
df = df_mycars[['driving', 'mpg']] #2. 연비 Table
df

driving  mpg
0         f   18
1         f   21
2         f   20
3         f   21
4         f   16
..      ...  ...
229       f   19
230       f   21
231       f   16
232       f   18
233       f   17

[234 rows x 2 columns]

In [50]:
count = df['driving'].value_counts().sort_index()
count

4    103
f    106
r     25
Name: driving, dtype: int64

In [51]:
cumcnt = np.cumsum(count)
cumcnt

4    103
f    209
r    234
Name: driving, dtype: int64

In [52]:
percent = count/sum(count)*100
percent

4    44.017094
f    45.299145
r    10.683761
Name: driving, dtype: float64

In [53]:
cumpct = np.cumsum(percent)
cumpct

4     44.017094
f     89.316239
r    100.000000
Name: driving, dtype: float64

In [54]:
count_data = pd.DataFrame({'Count': count, 'CumCnt':cumcnt,'Percent':percent, 'CumPct':cumpct})
count_data.columns.name='driving'
count_data

driving  Count  CumCnt    Percent      CumPct
4          103     103  44.017094   44.017094
f          106     209  45.299145   89.316239
r           25     234  10.683761  100.000000

# 기술통계 실습 3 (p. 230)
- 대한민국 남성의 20~30대 키를 측정한 결과, 평균 175cm, 표준편차 5cm이다.
- 나의 키가 180cm일 경우에 나는 몇 % 범위에 위치할까요?

In [55]:
#누적확률 값 계산
mu = 175
sigma = 5
x = 180
prob = stats.norm.cdf(x, mu, sigma) # 정규 분포의 누적 확률 값 계산
print('나의 키가 {}cm 일 경우에 나는 {:.2f}% 범위에 위치한다'.format(x, (1-prob)*100))

나의 키가 180cm 일 경우에 나는 15.87% 범위에 위치한다


# 기술통계 실습 4 (p. 233)
- 어떤 자동차 1리터당 주행거리가 평균 12km, 표준편차 3km인 정규분포를 이룬다.
1) 1리터를 가지고 9km 이상 15km 이하를 달릴 확률은?
2) 1리터를 가지고 15km 이상을 달릴 확률은?

In [56]:
mu = 12
sigma = 3
x = 15
prob = stats.norm.cdf(x, mu, sigma)
print('{}km 이상 {}km 이하를 달릴 확률은 {:.2f}이다.'.format(9, 15, (prob-0.5)*2))
print('{}km 이상 달릴 확률은 {:.2f}이다.'.format(15, (1-prob)))

9km 이상 15km 이하를 달릴 확률은 0.68이다.
15km 이상 달릴 확률은 0.16이다.


# 기술통계 실습 5 (p. 240)
- 확률변수 t는 자유도가 100인 t-분포에 따른다.
- t값이 2.0일 때 P(T <= t)를 계산하시오.

In [58]:
df = 100
t = 2
prob = stats.t.cdf(t,df)
print('P(T <= F): {:.3f}'.format(prob))

P(T <= F): 0.976


# 기술통계 실습 6 (p. 243)
- 확률변수 x^2는 자유도가 8인 x^2 분포를 따른다.
- x^2값이 3.0일 때 P(X <= x^2)를 계산하시오.

In [59]:
chisq = 3.0
df = 8
prob = stats.chi2.cdf(chisq,df)
print('P(X <= {}):{:.4f}'.format(chisq,prob))

P(X <= 3.0):0.0656


# 기술통계 실습 7 (p. 247)
- 확률변수 F는 각각 자유도가 10, 100인 F분포를 따른다.
- F값이 1.5일 때, P(X <= F)를 계산하시오.

In [60]:
f = 1.5
dfnum = 10
dfden = 100
prob = stats.f.cdf(f,dfnum,dfden)
print('P(X <= F): {:.3f}'.format(prob))

P(X <= F): 0.850
